### PR TITLE
Acom 6695 track sponsored article performance

### PR DESCRIPTION
**Pull request by helloaiway**
## Jira
Acceptance Criteria
1. Add tracking to ALL pieces of content articles.
2. Mitchell needs to sign off this story before release to production.

## Dependency
This is dependent by https://gitlab.dev.activenetwork.com/activenetwork/a3/merge_requests/526

## Test
http://a3coreint12.dev.activenetwork.com/triathlon/articles/holiday-gift-guide-for-triathletes-2

### CODE REVIEW

**Comment by MaikZhang** [comment]
Please also bounce the ACTV version for this.

File: lib/actv/asset.rb [hunk 1]
```diff
@@ -403,5 +412,11 @@ def kids_interest?
       interests = meta_interests.to_a.map(&:downcase)
       ['kids', 'family'].any? { |tag| interests.include? tag }
     end
+
+    def sponsored_date_available?
+      start_time = Time.parse(sponsoredContent[:startDate])
+      end_time = Time.parse(sponsoredContent[:endDate])
+      start_time < Time.now && Time.now < end_time
```
Comment (helloaiway) [inline lib/actv/asset.rb]: cool

File: lib/actv/asset.rb [hunk 2]
```diff
@@ -22,7 +22,7 @@ class Asset < ACTV::Identity
 
     attr_reader :assetGuid, :assetName, :assetDsc, :activityStartDate, :activityStartTime, :activityEndDate, :activityEndTime,
       :homePageUrlAdr, :isRecurring, :contactName, :contactEmailAdr, :contactPhone, :showContact, :publishDate, :createdDate, :modifiedDate,
-      :authorName, :is_event, :is_article, :currencyCd, :contactTxt, :regReqMinAge, :regReqMaxAge, :regReqGenderCd
+      :authorName, :is_event, :is_article, :currencyCd, :contactTxt, :regReqMinAge, :regReqMaxAge, :regReqGenderCd, :sponsoredContent
```
Comment (MarkWang2) [inline lib/actv/asset.rb]: We do not need to guess the unpredictable future.

File: spec/actv/asset_spec.rb [hunk 3]
```diff
@@ -556,4 +557,48 @@ def self.env
...
+    let(:response) { { assetGuid: 1 } }
+    let(:asset) { ACTV::Asset.new response }
+
+    context 'when sponsored content is nil' do
+      it 'returns nil' do
+        expect(asset.sponsored_content).to be_nil
+      end
+    end
+
+    context 'when sponsored content is exist' do
```
Comment (hlcfan) [inline spec/actv/asset_spec.rb]: "when sponsored content exists"

File: spec/actv/asset_spec.rb [hunk 4]
```diff
@@ -556,4 +557,48 @@ def self.env
...
+        expect(asset.sponsored_content).to be_a Hash
+      end
+    end
+  end
+
+  describe '#sponsored?' do
+    let(:response) { { assetGuid: 1 } }
+    let(:asset) { ACTV::Asset.new response }
+
+    context 'when sponsored status is unenabled' do
```
Comment (hlcfan) [inline spec/actv/asset_spec.rb]: "when sponsored status is disabled"

File: spec/actv/asset_spec.rb (hunk 3 above)
Comment (helloaiway) [inline spec/actv/asset_spec.rb]: cool

File: spec/actv/asset_spec.rb (hunk 4 above)
Comment (helloaiway) [inline spec/actv/asset_spec.rb]: cool

File: spec/actv/asset_spec.rb [hunk 5]
```diff
@@ -556,4 +557,48 @@ def self.env
...
+      let(:response) { { assetGuid: 1, sponsoredContent: { startDate: '2016-12-15T00:00:00', enabled: 'false', endDate: '2017-10-28T11:59:59' } } }
+      it 'returns false' do
+        expect(asset.sponsored?).to be_false
+      end
+    end
+
+    context 'when sponsored date is valaid and status is enabled' do
+      let(:response) { { assetGuid: 1, sponsoredContent: { startDate: '2016-12-15T00:00:00', enabled: 'true', endDate: Time.at(Time.now.to_i + 3600*24).strftime("%Y-%m-%dT%H:%I:%S") } } }
```
Comment (hlcfan) [inline spec/actv/asset_spec.rb]: `Time.at(Time.now.to_i + 3600*24).strftime("%Y-%m-%dT%H:%I:%S")` can be 
`Time.now + 1.day` or `DateTime.tomorrow`

File: lib/actv/asset.rb (hunk 2 above)
Comment (helloaiway) [inline lib/actv/asset.rb]: ok. removed.

File: spec/actv/asset_spec.rb (hunk 5 above)
Comment (helloaiway) [inline spec/actv/asset_spec.rb]: No the tomorrow method isn't exist since it's doesn't use rails.

File: lib/actv/asset.rb [hunk 6]
```diff
@@ -354,6 +355,14 @@ def organization
       @attrs[:organization] || {}
     end
 
+    def sponsored?
+      sponsoredContent.present? && sponsoredContent[:enabled].to_s == 'true' && sponsored_date_available?
+    end
+
+    def sponsoredContent
+      @sponsoredContent ||= @attrs[:sponsoredContent] unless @attrs[:sponsoredContent].nil?
```
Comment (helloaiway) [inline lib/actv/asset.rb]: Maybe, But i think we don't always trust as data. I thinks this is more safe.

File: spec/actv/asset_spec.rb [hunk 7]
```diff
@@ -556,4 +557,30 @@ def self.env
...
+      it 'returns true' do
+        expect(asset.sponsored?).to be_true
+      end
+    end
+
+    context 'when sponsored date is invalid and status is enabled' do
+      let(:response) { { assetGuid: 1, sponsoredContent: { startDate: '2015-12-15T00:00:00', enabled: 'true', endDate: '2016-10-28T11:59:59' } } }
+      it 'returns false' do
+        expect(asset.sponsored?).to be_false
```
Comment (RogerTongTong) [inline spec/actv/asset_spec.rb]: test

File: spec/actv/asset_spec.rb [hunk 8]
```diff
@@ -556,4 +557,30 @@ def self.env
...
+
+    context 'when sponsored status is disabled' do
+      let(:response) { { assetGuid: 1, sponsoredContent: { startDate: '2016-12-15T00:00:00', enabled: 'false', endDate: '2017-10-28T11:59:59' } } }
+      it 'returns false' do
+        expect(asset.sponsored?).to be_false
+      end
+    end
+
+    context 'when sponsored date is valid and status is disabled' do
+      let(:response) { { assetGuid: 1, sponsoredContent: { startDate: '2016-12-15T00:00:00', enabled: 'true', endDate: Time.at(Time.now.to_i + 3600*24).strftime("%Y-%m-%dT%H:%I:%S") } } }
```
Comment (helloaiway) [inline spec/actv/asset_spec.rb]: Nice, Have updated.Thanks.

File: spec/actv/asset_spec.rb [hunk 9]
```diff
@@ -537,23 +538,60 @@ def self.env
...
+      let(:response) { { assetGuid: 1, sponsoredContent: { startDate: '2016-12-15T00:00:00', enabled: 'false', endDate: '2017-10-28T11:59:59' } } }
+      it 'returns false' do
+        expect(asset.sponsored?).to be_false
+      end
+    end
+
+    context 'when sponsored date is valid and status is disabled' do
+      before do
+        Timecop.freeze(2017,1,1,0,0,0)
+      end
```
Comment (MarkWang2) [inline spec/actv/asset_spec.rb]: before { Timecop.freeze(2017,1,1,0,0,0) }

File: spec/actv/asset_spec.rb [hunk 10]
```diff
@@ -537,23 +538,60 @@ def self.env
...
+
+      after do
+        Timecop.return
+      end
```
Comment (MarkWang2) [inline spec/actv/asset_spec.rb]: after { Timecop.return }

Comment (helloaiway) [inline spec/actv/asset_spec.rb]: cool

File: spec/actv/asset_spec.rb (hunk 9 above)
Comment (helloaiway) [inline spec/actv/asset_spec.rb]: cool

**Comment by RogerTongTong** [comment]
Looks great. Thanks @helloaiway

**Comment by RogerTongTong** [comment]
@hlcfan I suppose we could merge this MR and publish to RubyGem.

**Comment by hlcfan** [comment]
🚢

**Pull request merged**